# Imports

In [1]:
import pandas as pd
pd.set_option('display.max_columns', None)
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
raw_data = pd.read_csv('hotel_bookings.csv')
raw_data.sample(8)

hotel  is_canceled  lead_time  arrival_date_year  \
29499   Resort Hotel            0         17               2016   
71818     City Hotel            1        414               2017   
82322     City Hotel            0         35               2015   
106146    City Hotel            0          3               2017   
67011     City Hotel            1        111               2017   
51281     City Hotel            1         64               2016   
43748     City Hotel            0         86               2015   
113754    City Hotel            0        130               2017   

       arrival_date_month  arrival_date_week_number  \
29499             October                        44   
71818                July                        28   
82322            December                        53   
106146           February                         8   
67011               April                        17   
51281                 May                        21   
43748           September                        39   
113754               June                        23   

        arrival_date_day_of_month  stays_in_weekend_nights  \
29499                          27                        0   
71818                          13                        0   
82322                          28                        1   
106146                         21                        0   
67011                          27                        1   
51281                          17                        0   
43748                          24                        0   
113754                          8                        0   

        stays_in_week_nights  adults  children  babies meal country  \
29499                      3       2       0.0       0   HB     ESP   
71818                      2       2       0.0       0   BB     PRT   
82322                      0       2       2.0       0   SC     PRT   
106146                     1       2       0.0       0   BB     PRT   
67011                      3       2       0.0       0   BB     PRT   
51281                      2       2       0.0       0   BB     PRT   
43748                      2       1       0.0       0   BB     PRT   
113754                     3       3       0.0       0   BB     GBR   

       market_segment distribution_channel  is_repeated_guest  \
29499       Online TA                TA/TO                  0   
71818   Offline TA/TO                TA/TO                  0   
82322       Online TA                TA/TO                  0   
106146         Direct               Direct                  0   
67011   Offline TA/TO                TA/TO                  0   
51281       Corporate                TA/TO                  0   
43748   Offline TA/TO                TA/TO                  0   
113754      Online TA                TA/TO                  0   

        previous_cancellations  previous_bookings_not_canceled  \
29499                        0                               0   
71818                        0                               0   
82322                        0                               0   
106146                       0                               0   
67011                        0                               0   
51281                        0                               0   
43748                        0                               0   
113754                       0                               0   

       reserved_room_type assigned_room_type  booking_changes deposit_type  \
29499                   D                  E                0   No Deposit   
71818                   A                  B                1   No Deposit   
82322                   F                  F                0   No Deposit   
106146                  A                  A                1   No Deposit   
67011                   A                  A                0   Non Refund   
51281                   A                  B                1   No 

## 1. Data Collection


In [3]:
raw_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

- There are 32 columns in the dataset.
- The dataset contains 119,390 rows.
- Data Types:
    - 1. 16 columns are of int64 data type.
    - 2. 4 columns are of float64 data type.
    - 3. 12 columns are of object data types. 
- Missing Values:
    - The column `children` has <span style="color:red">4</span> missing values.
    - The column `country` has <span style="color:red">488</span> missing values.
    - The column `agent` has <span style="color:red">16,340</span> missing values.
    - The column `company` has <span style="color:red">112,593</span> missing values.


## 2 Exploratory Data Analysis (EDA)

### 2.1 Looking for missing values.

In [4]:
raw_data.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

- Dropping null values is easy but isn't a good practice, so we will start with replacing them with zero.

In [5]:
raw_data.fillna(0, inplace=True)

### 2.2 cleaning Data

In [6]:
raw_data.describe().T

count         mean         std      min  \
is_canceled                     119390.0     0.370416    0.482918     0.00   
lead_time                       119390.0   104.011416  106.863097     0.00   
arrival_date_year               119390.0  2016.156554    0.707476  2015.00   
arrival_date_week_number        119390.0    27.165173   13.605138     1.00   
arrival_date_day_of_month       119390.0    15.798241    8.780829     1.00   
stays_in_weekend_nights         119390.0     0.927599    0.998613     0.00   
stays_in_week_nights            119390.0     2.500302    1.908286     0.00   
adults                          119390.0     1.856403    0.579261     0.00   
children                        119390.0     0.103886    0.398555     0.00   
babies                          119390.0     0.007949    0.097436     0.00   
is_repeated_guest               119390.0     0.031912    0.175767     0.00   
previous_cancellations          119390.0     0.087118    0.844336     0.00   
previous_bookings_not_canceled  119390.0     0.137097    1.497437     0.00   
booking_changes                 119390.0     0.221124    0.652306     0.00   
agent                           119390.0    74.828319  107.141953     0.00   
company                         119390.0    10.775157   53.943884     0.00   
days_in_waiting_list            119390.0     2.321149   17.594721     0.00   
adr                             119390.0   101.831122   50.535790    -6.38   
required_car_parking_spaces     119390.0     0.062518    0.245291     0.00   
total_of_special_requests       119390.0     0.571363    0.792798     0.00   

                                    25%       50%     75%     max  
is_canceled                        0.00     0.000     1.0     1.0  
lead_time                         18.00    69.000   160.0   737.0  
arrival_date_year               2016.00  2016.000  2017.0  2017.0  
arrival_date_week_number          16.00    28.000    38.0    53.0  
arrival_date_day_of_month          8.00    16.000    23.0    31.0  
stays_in_weekend_nights            0.00     1.000     2.0    19.0  
stays_in_week_nights               1.00     2.000     3.0    50.0  
adults                             2.00     2.000     2.0    55.0  
children                           0.00     0.000     0.0    10.0  
babies                             0.00     0.000     0.0    10.0  
is_repeated_guest                  0.00     0.000     0.0     1.0  
previous_cancellations             0.00     0.000     0.0    26.0  
previous_bookings_not_canceled     0.00     0.000     0.0    72.0  
booking_changes                    0.00     0.000     0.0    21.0  
agent                              7.00     9.000   152.0   535.0  
company                            0.00     0.000     0.0   543.0  
days_in_waiting_list               0.00     0.000     0.0   391.0  
adr                               69.29    94.575   126.0  5400.0  
required_car_parking_spaces        0.00     0.000     0.0     8.0  
total_of_special_requests          0.00     0.000     1.0     5.0

In [7]:
# Display rows with negative adr values
negative_adr_rows = raw_data[raw_data['adr'] < 0]
display(negative_adr_rows)

# Drop rows with negative adr values
raw_data.drop(negative_adr_rows.index, inplace=True)

hotel  is_canceled  lead_time  arrival_date_year  \
14969  Resort Hotel            0        195               2017   

      arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
14969              March                        10                          5   

       stays_in_weekend_nights  stays_in_week_nights  adults  children  \
14969                        4                     6       2       0.0   

       babies meal country market_segment distribution_channel  \
14969       0   BB     GBR         Groups               Direct   

       is_repeated_guest  previous_cancellations  \
14969                  1                       0   

       previous_bookings_not_canceled reserved_room_type assigned_room_type  \
14969                               2                  A                  H   

       booking_changes deposit_type  agent  company  days_in_waiting_list  \
14969                2   No Deposit  273.0      0.0                     0   

         customer_type   adr  required_car_parking_spaces  \
14969  Transient-Party -6.38                            0   

       total_of_special_requests reservation_status reservation_status_date  
14969                          0          Check-Out                15-03-17

### Based on the following logic:

1.  Rows containing negative values in the `adr` column.
2.  The `distribution_channel` column, as it depicts the same information as the `market_segment` column.
3.  The `arrival_date_week_number` column, as the week number can be derived from the `arrival_date` column.
4.  Rows where both the `children` and `adults` columns contain zero.

In [8]:
def clean_dataframe(data):
    """
    Cleans a pandas DataFrame based on specified criteria.

    Args:
        raw_data: The pandas DataFrame to clean.

    Returns:
        The cleaned pandas DataFrame.
    """

    # 1. Drop rows with negative 'adr' values.
    data.drop(data[data['adr'] <= 0].index, inplace=True)

    # 2. Drop the 'distribution_channel' column.
    if 'distribution_channel' in data.columns: # check if column exists
        data = data.drop(columns=['distribution_channel'])

    # 3. Drop the 'arrival_date_week_number' column.
    if 'arrival_date_week_number' in data.columns: # check if column exists
        data = data.drop(columns=['arrival_date_week_number'])

    # 4. Drop rows where both 'children' and 'adults' are zero.
    data = data[~((data['children'] == 0) & (data['adults'] == 0))] # ~ is the not operator.

    return data


raw_data = clean_dataframe(raw_data)

In [9]:
raw_data

hotel  is_canceled  lead_time  arrival_date_year  \
2       Resort Hotel            0          7               2015   
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
5       Resort Hotel            0         14               2015   
6       Resort Hotel            0          0               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0         23               2017   
119386    City Hotel            0        102               2017   
119387    City Hotel            0         34               2017   
119388    City Hotel            0        109               2017   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_day_of_month  stays_in_weekend_nights  \
2                    July                          1                        0   
3                    July                          1                        0   
4                    July                          1                        0   
5                    July                          1                        0   
6                    July                          1                        0   
...                   ...                        ...                      ...   
119385             August                         30                        2   
119386             August                         31                        2   
119387             August                         31                        2   
119388             August                         31                        2   
119389             August                         29                        2   

        stays_in_week_nights  adults  children  babies meal country  \
2                          1       1       0.0       0   BB     GBR   
3                          1       1       0.0       0   BB     GBR   
4                          2       2       0.0       0   BB     GBR   
5                          2       2       0.0       0   BB     GBR   
6                          2       2       0.0       0   BB     PRT   
...                      ...     ...       ...     ...  ...     ...   
119385                     5       2       0.0       0   BB     BEL   
119386                     5       3       0.0       0   BB     FRA   
119387                     5       2       0.0       0   BB     DEU   
119388                     5       2       0.0       0   BB     GBR   
119389                     7       2       0.0       0   HB     DEU   

       market_segment  is_repeated_guest  previous_cancellations  \
2              Direct                  0                       0   
3           Corporate                  0                       0   
4           Online TA                  0                       0   
5           Online TA                  0                       0   
6              Direct                  0                       0   
...               ...                ...                     ...   
119385  Offline TA/TO                  0                       0   
119386      Online TA                  0                       0   
119387      Online TA                  0                       0   
119388      Online TA                  0                       0   
119389      Online TA                  0                       0   

        previous_bookings_not_canceled reserved_room_type assigned_room_type  \
2                                    0                  A                  C   
3                                    0                  A                  A   
4                                    0                  A                  A   
5                                    0                  A                  A   
6                                    0                  C                  C   
...                                ...                ...                ...   
119385 

### 2.3 Feature Selection

##### Why Feature Selection Might Be Beneficial:

* **Improved LLM Performance:** LLMs can be sensitive to irrelevant or redundant information. Feature selection can help focus the LLM on the most important data points, potentially improving accuracy and relevance.

* **Reduced Dimensionality:** If your dataset has a large number of features, feature selection can reduce the dimensionality of the data. This can lead to:

    1.  Faster vector embedding generation and retrieval.
    2.  Reduced storage requirements for your vector database.
    3.  Improved efficiency of your API.

In [10]:
raw_data['booking_source'] = np.where(
    (raw_data['agent'] > 0) & (raw_data['agent'].notna()), 
    'agent',
    np.where(
        (raw_data['company'] > 0) & (raw_data['company'].notna()),
        'company',
        'self'
    )
)

# Enhanced customer history features (FIXED VERSION)
raw_data['previous_experience'] = np.where(
    raw_data['is_repeated_guest'] == 1,
    'Repeat',
    'First-time'
)

# For previous cancellations analysis
raw_data['previous_cancel_behavior'] = np.where(
    raw_data['previous_cancellations'] > 0,
    'Previous Canceller',
    'No Cancel History'
)

# Create customer segments
conditions = [
    raw_data['previous_cancellations'] > 0,
    raw_data['is_repeated_guest'] == 1,
    raw_data['total_of_special_requests'] > 2
]
choices = ['Risk', 'Loyal', 'High-Maintenance']
raw_data['customer_segment'] = np.select(conditions, choices, default='Standard')

raw_data['arrival_date'] = pd.to_datetime(
    raw_data['arrival_date_year'].astype(str) + '-' +
    raw_data['arrival_date_month'] + '-' +
    raw_data['arrival_date_day_of_month'].astype(str)
)

columns_to_drop = [
    'arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month',
    'assigned_room_type','days_in_waiting_list', 'reservation_status_date',
    'agent', 'company','is_repeated_guest', 'previous_cancellations','reservation_status'
]
raw_data.drop(columns=columns_to_drop, inplace=True)

- Combinied the `agent` and `company` columns into a single `booking_source` column.
- Determines whether a booking was made through an **agent**, a **company**, or directly by the customer (**self**).

- Created a new column `previous_experience` to categorise customers as either **Repeat** or **First-time** guests.
- Created a new column `previous_cancel_behavior` to categorise customers based on their cancellation history.
- Combined `arrival_date_year`, `arrival_date_month`, and `arrival_date_day_of_month` into a single `arrival_date` column.
- Removed redundant or unnecessary columns from the dataset, to focuses on features that are most relevant for analytics and chatbot queries.


In [11]:
raw_data.to_csv('cleaned_hotel_bookings.csv', index=False)

### 2.4 Chatbot-Focused Text Representation

In [12]:
def create_chatbot_entry(row):
    # Format dates
    arrival_date = row['arrival_date'].strftime('%b %d, %Y')
    
    # Convert codes to human-readable formats
    meal_map = {
        'BB': 'Bed & Breakfast',
        'HB': 'Half Board',
        'FB': 'Full Board',
        'SC': 'Self Catering'
    }
    
    customer_type_map = {
        'Transient': 'Individual Traveler',
        'Contract': 'Corporate Contract',
        'Group': 'Tour Group',
        'Transient-Party': 'Travel Group'
    }
    
    return f"""
    {row['hotel']} booking {'CANCELED' if row['is_canceled'] else 'CONFIRMED'}.
    - Dates: Arriving {arrival_date} for {row['stays_in_weekend_nights']+row['stays_in_week_nights']} nights
    - Guests: {row['adults']} adults, {row['children']} children
    - Room: {row['reserved_room_type']} (Booked), {row.get('assigned_room_type', row['reserved_room_type'])} (Assigned)
    - Pricing: ${row['adr']:.2f} per night (Total: ${row['adr']*(row['stays_in_weekend_nights']+row['stays_in_week_nights']):.2f})
    - Special Requests: {row['total_of_special_requests']} ({'No parking needed' if row['required_car_parking_spaces'] == 0 else 'Parking required'})
    - Customer Profile: {customer_type_map.get(row['customer_type'], row['customer_type'])} | {row['previous_experience']} guest
    - Booking Source: {row['booking_source']} via {row['market_segment']}
    - Risk Factors: {row['previous_cancel_behavior']} | {row['customer_segment']} segment
    - Meal Plan: {meal_map.get(row['meal'], row['meal'])}
    - Country: {row['country']}
    """

raw_data['chatbot_text'] = raw_data.apply(create_chatbot_entry, axis=1)